### PR TITLE
Add time function constants to temporal algebra

### PR DESCRIPTION
**Pull request by ninsbl**
This PR adds a parser rule for time functions that return a numerical constant for a map (like `start_doy`) in the temporal algebra. This allows to use e.g. the `start_doy` from the time stamp of a map as a constant in temporal algebra expressions in `t.rast.algebra`. 

Example:
`t.rast.algebra -d expression="R = if(A >= 3, start_doy(A, 0), 0)" basename="r" suffix="gran"`

Addresses #891

### CODE REVIEW

**Comment by ninsbl** [comment]
This PR adds a version of time functions returning numeric values for elements of a time series.
For this function type a neighborhood modifier is added that a) distinguishes it from the function used for temporal selection (for the lexical analyser) and b) adds more options for usage.

The function looks e.g. like this:
start_doy(A, -1)

If that is an acceptable solution, I will add to the documentation before evtl. merging the PR.

**Comment by veroandreo** [comment]
I've tested this PR, but I still get errors for the originally reported problem in #891 

```
 t.rast.algebra -d basename=r expression="R=start_doy(cla)"

Traceback (most recent call last):
  File "/home/veroandreo/software/grass79-dev/dist.x86_64-pc-linux-gnu/scripts/t.rast.algebra", line 133, in <module>
    sys.exit(main())
  File "/home/veroandreo/software/grass79-dev/dist.x86_64-pc-linux-gnu/scripts/t.rast.algebra", line 125, in main
    pc = p.parse(expression, basename, grass.script.overwrite())
  File "/home/veroandreo/software/grass79-dev/dist.x86_64-pc-linux-gnu/etc/python/grass/temporal/temporal_raster_algebra.py", line 111, in parse
    self.parser.parse(expression)
  File "/usr/lib/python3.8/site-packages/ply/yacc.py", line 333, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/usr/lib/python3.8/site-packages/ply/yacc.py", line 1201, in parseopt_notrack
    tok = call_errorfunc(self.errorfunc, errtoken, self)
  File "/usr/lib/python3.8/site-packages/ply/yacc.py", line 192, in call_errorfunc
    r = errorfunc(token)
  File "/home/veroandreo/software/grass79-dev/dist.x86_64-pc-linux-gnu/etc/python/grass/temporal/temporal_algebra.py", line 3089, in p_error
    raise SyntaxError("syntax error on line %d, position %i token %s near '%s' expression '%s'" %
SyntaxError: syntax error on line 1, position 15 token RPAREN near ')' expression 'R=start_doy(cla)'
```
`cla` is a time series of modis chlorophyll data:

```
t.info cla
 +-------------------- Space Time Raster Dataset -----------------------------+
 |                                                                            |
 +-------------------- Basic information -------------------------------------+
 | Id: ........................ cla@pruebas
 | Name: ...................... cla
 | Mapset: .................... pruebas
 | Creator: ................... veroandreo
 | Temporal type: ............. absolute
 | Creation time: ............. 2017-06-11 19:59:51.590922
 | Modification time:.......... 2017-06-11 20:00:36.812285
 | Semantic type:.............. mean
 +-------------------- Absolute time -----------------------------------------+
 | Start time:................. 2003-01-01 00:00:00
 | End time:................... 2014-01-01 00:00:00
 | Granularity:................ 1 day
 | Temporal type of maps:...... interval
 +-------------------- Spatial extent ----------------------------------------+
 | North:...................... -38.0
 | South:...................... -55.0
 | East:.. .................... -55.0
 | West:....................... -70.0
 | Top:........................ 0.0
 | Bottom:..................... 0.0
 +-------------------- Metadata information ----------------------------------+
 | Raster register table:...... raster_map_register_600c2c6126544008a595bc8a3f97fad1
 | North-South resolution min:. 0.041667
 | North-South resolution max:. 0.041667
 | East-west resolution min:... 0.041667
 | East-west resolution max:... 0.041667
 | Minimum value min:.......... 0.02925
 | Minimum value max:.......... 0.26472
 | Maximum value min:.......... 5.2104
 | Maximum value max:.......... 99.953934
 | Aggregation type:........... None
 | Number of registered bands:. None
 | Number of registered maps:.. 460
```

**Comment by veroandreo** [comment]
I have also tested the new behavior, and I get a lot of things printed to stdout. Is that the intended output? Or what is the right way to test this indeed new feature?

```
t.rast.algebra -d basename=r expression="R=start_doy(cla, -1)" | less

1
9
17
25
33
41
49
57
65
73

... >> only a sample is shown

1
9
17
25
33
41
49
57
65
73

...>> only a sample is shown

{'STDS': {'name': 'R', 'stdstype': 'strds', 'temporal_type': 'absolute'},
 'processes': [{'flags': [],
                'inputs': [('expression', 'r_000=1'), ('region', 'current')],
                'name': 'r.mapcalc',
                'outputs': []},
               {'flags': [],
                'inputs': [('expression', 'r_001=9'), ('region', 'current')],
                'name': 'r.mapcalc',
                'outputs': []},
               {'flags': [],
                'inputs': [('expression', 'r_002=17'), ('region', 'current')],
                'name': 'r.mapcalc',
                'outputs': []},
               {'flags': [],
                'inputs': [('expression', 'r_003=25'), ('region', 'current')],
                'name': 'r.mapcalc',
                'outputs': []},
               {'flags': [],
                'inputs': [('expression', 'r_004=33'), ('region', 'current')],
                'name': 'r.mapcalc',
                'outputs': []},

...>> only a sample is shown

 'register': [('r_000', '2003-01-09 00:00:00', '2003-01-17 00:00:00'),
              ('r_001', '2003-01-17 00:00:00', '2003-01-25 00:00:00'),
              ('r_002', '2003-01-25 00:00:00', '2003-02-02 00:00:00'),
              ('r_003', '2003-02-02 00:00:00', '2003-02-10 00:00:00'),
              ('r_004', '2003-02-10 00:00:00', '2003-02-18 00:00:00'),
              ('r_005', '2003-02-18 00:00:00', '2003-02-26 00:00:00'),
              ('r_006', '2003-02-26 00:00:00', '2003-03-06 00:00:00'),
              ('r_007', '2003-03-06 00:00:00', '2003-03-14 00:00:00'),
              ('r_008', '2003-03-14 00:00:00', '2003-03-22 00:00:00'),

...>> only a sample is shown
```

**Comment by ninsbl** [comment]
Thanks for testing, Vero.

Yes. When I tried to stick with the original function, conflicts arose if used for temporal selection.
Therefore I added a neighborhood modifier to the function that returns a numeric constant in the r.mapcalc expression. 
`start_doy(cla, -1)` f.ex. returns the start doy of the previous map.

So, this should work:
`t.rast.algebra -d basename=r expression="R=if(cla>12.0&&cla[-1]>12.0&&cla[-2]>12.0, start_doy(cla, 0), 0)"`

**Comment by ninsbl** [comment]
Thanks also for the heads up with the console output. I forgot to remove a print statement. Fixed that now.

**Comment by veroandreo** [comment]
I now removed the former patch, and applied this one. However, I still get a lot of things printed in stdout, no error apparently, but no new STRDS is created...

```
t.rast.algebra -d basename=r expression="R=if(cla>5.0 && cla[-1]>5.0 && cla[-2]>5.0, start_doy(cla,0), 0)"
GRASS 7.9.dev (latlong_wgs84):~ > t.info R
ERROR: Dataset <R@pruebas> not found in temporal database
```

**Comment by ninsbl** [comment]
> Thanks for testing, Vero.
> 
> Yes. When I tried to stick with the original function, conflicts arose if used for temporal selection.
> Therefore I added a neighborhood modifier to the function that returns a numeric constant in the r.mapcalc expression.
> `start_doy(cla, -1)` f.ex. returns the start doy of the previous map.
> 
> So, this should work:
> `t.rast.algebra -d basename=r expression="R=if(cla>12.0&&cla[-1]>12.0&&cla[-2]>12.0, start_doy(cla, 0), 0)"`

Yes, because of the d-flag (dry run). It prints out the generated r.mapcalc commands (amongst others), which is handy for testing, but does not generate any data...
Clould be less chatty if not in debug mode though...

**Comment by veroandreo** [comment]
> > Thanks for testing, Vero.
> > Yes. When I tried to stick with the original function, conflicts arose if used for temporal selection.
> > Therefore I added a neighborhood modifier to the function that returns a numeric constant in the r.mapcalc expression.
> > `start_doy(cla, -1)` f.ex. returns the start doy of the previous map.
> > So, this should work:
> > `t.rast.algebra -d basename=r expression="R=if(cla>12.0&&cla[-1]>12.0&&cla[-2]>12.0, start_doy(cla, 0), 0)"`
> 
> Yes, because of the d-flag (dry run). It prints out the generated r.mapcalc commands (amongst others), which is handy for testing, but does not generate any data...
> Clould be less chatty if not in debug mode though...

ops... my bad! Yes, the example works fine :)

**Comment by ninsbl** [comment]
If there are no objections, this should be actually ready to merge.

I would open a new PR for the documentation of the function in the algebra module (https://grass.osgeo.org/grass79/manuals/t.rast.algebra.html).

**Comment by ninsbl** [comment]
When updating the documentation I will also adress: #1065

File: temporal/t.rast.algebra/t.rast.algebra.html [hunk 1]
```diff
@@ -274,6 +274,11 @@ <h4>Temporal functions</h4>
 end_second(A)            The second of the end time [0 - 59]
 </pre></div>
 
+In order to use the numbers returned by the functios in the last block above, an
```
Comment (neteler) [inline temporal/t.rast.algebra/t.rast.algebra.html:277]:
```suggestion
In order to use the numbers returned by the functions in the last block above, an 
```

File: temporal/t.rast.algebra/t.rast.algebra.html [hunk 2]
```diff
@@ -561,6 +567,13 @@ <h3>Computation with time intervals with temporal topology relation</h3>
 C = if({equal}, B {#,contain} A > 1, (B {+,contain,l} A {-,equal,l} B) {equal,=/} (B {#,contain} A), A)
 </pre></div>
 
+<h3>Compute DOY for spatio-temporal conditions</h3>
+Compute the DOY for all maps from STRDS A where conditions are met at three 
+consecutive time intervals (e.g. temperature > 0):
+<div class="code"><pre>
+B = if(A>0.0&&A[-1]>0.0&&A[-2]>0.0, start_doy(A, -1), 0)"
```
Comment (veroandreo) [inline temporal/t.rast.algebra/t.rast.algebra.html:574]:
```suggestion
B = if(A > 0.0 && A[-1] > 0.0 && A[-2] > 0.0, start_doy(A, -1), 0)"
```
adding spaces for better readability

**Comment by veroandreo** [comment]
@ninsbl what is missing here for merging?

**Comment by ninsbl** [comment]
Messed up and replaced by: #2835

**Comment by wenzeslaus** [comment]
> Messed up...

It is no tragedy that a very old PR is replaced by a new one, but for the sake of increasing our collective Git skills, here is my comment:

I actually don't know how exactly this happens and I would be interested in knowing that. Anyway, the rebase seems to be done sort of backwards. The commits from main are on this branch, but as different commits (they have different hash):

https://github.com/OSGeo/grass/pull/905/commits/b9cb2713390c0494503c6cb95d59b76d1808777f
https://github.com/OSGeo/grass/commit/1f2b44849905808ef5553a56f30af5165cae8d1b

That's what is supposed to happen during rebase, but with commits on your feature branch, not on the main branch.

The following will reapply commits from my-feature-branch to upstream/main. Good.

```
git switch my-feature-branch
git rebase upstream/main
```

The following will reapply commits from main to my-feature-branch. Bad.

```
git switch main
git rebase my-feature-branch
```

I actually never tried to do rebase after I did merge on that branch because it sounds like something I don't want to do. [This SO answer](https://stackoverflow.com/a/17141512/574907) with `git merge --squash` looks hopeful, though. The other way around, merge after rebase, is not a problem.

The rebase commands are the same as merge commands.

The following will merge upstream/main into my-feature-branch which is exactly what the automatic commit message says:

```
git switch my-feature-branch
git merge upstream/main
```

The upstream/main stays as is (and so would main if you use that), but my-feature-branch now has one more commit which is the merge commit. (Analogous to reapplied commits on my-feature-branch when rebasing.)

**Comment by neteler** [comment]
> It is no tragedy that a very old PR is replaced by a new one

Surely no tragedy but contributions by others are lost if "simply" a new PR is opened.
So hopefully this one can be cleaned up.

**Comment by ninsbl** [comment]
Good points. I don`t have the history of yesterdays changes unfortunately, so I cannot say exactly how exactly i messed up.
Now I did:
```
# 1) a hard reset to the last reasonable commit, renamed the branch temporarily, created
git reset --hard 83be7ee867b0274ebf4f3ae44f0fd57faddc6276
# 2) temporarily rename the PR-branch 
git branch -m temporal_algebra_constants-old
# 3) Switch to master
git checkout main
# 4) Get latest changes from upstream
git fetch --all
git rebase upstream/main
# 5) Create new branch from latest main branch
git checkout -b temporal_algebra_constants
# 6) Merge changes from the PR (temporarily renamed branch) into new, clean branch from main
git merge temporal_algebra_constants-old
# 7) Temporarily deactivate pre-commit hooks as they were interfering with the merge process
pre-commit uninstall
# 8) Solve evtl. merge conflicts and finish merge
git add temporal/t.rast.algebra/t.rast.algebra.html
git merge --continue
# 9) Reactivate pre-commit hooks
pre-commit install
# 10) Commit changes e.g. from re-applying pre-commit hooks
git commit -m "merge" temporal/t.rast.algebra/t.rast.algebra.html
# 11) Push the result to git to update the PR
git push origin temporal_algebra_constants
```
Seemed to have worked OK and preserved the history in the PR...

**Comment by wenzeslaus** [comment]
Brave!

**Comment by ninsbl** [comment]
OK to merge into main? Or should we wait for 8.4?

**Comment by ninsbl** [comment]
Adresses also #1065

File: temporal/t.rast.algebra/t.rast.algebra.html [hunk 3]
```diff
@@ -445,7 +450,8 @@ <h3>Temporal neighbourhood modifier</h3>
 The neighbourhood modifier of <em>r.mapcalc</em> is extended for the temporal
 raster algebra with the temporal dimension. The format is strds[t,r,c],
 where t is the temporal offset, r is the row offset and c is the column
-offset.
+offset. A single neighborhood modifier is interpreted as temporal offset [t],
+while two neighborhood modifier are interpreted as row and column offset [r,c].
```
Comment (veroandreo) [inline temporal/t.rast.algebra/t.rast.algebra.html:454]:
```suggestion
while two neighborhood modifiers are interpreted as row and column offsets [r,c].
```

**Comment by veroandreo** [comment]
> OK to merge into main? Or should we wait for 8.4?

I'd like to see it merged for 8.3, it was working fine already 2 and half years ago :)

**Review by wenzeslaus — COMMENTED** [review]
There is an example now and a low-level test. Surrounding code has tests, too. I would prefer test for resulting functionality on user-level, i.e., one for t.rast.algebra or equivalent. t.rast.algebra already has gunittest-based tests. pytest-based test would work, too.

File: temporal/t.rast.algebra/testsuite/test_raster_algebra.py [hunk 4]
```diff
@@ -258,6 +259,51 @@ def test_simple_arith_td_4(self):
...
+        ref_str = "...".join(
+            [
+                "r_2001_01_01=if(a1@... >= 3, 1, 0)",
+                "r_2001_01_02=if(a2@... >= 3, 2, 0)",
+                "r_2001_01_03=if(a3@... >= 3, 3, 0)",
+                "r_2001_01_04=if(a4@... >= 3, 4, 0)",
+            ]
+        )
+        ref_str = f"...{ref_str}..."
+        self.assertLooksLike(str(E.outputs.stdout), ref_str)
```
Comment (wenzeslaus) [inline temporal/t.rast.algebra/testsuite/test_raster_algebra.py:292]: My guess would be that this is an implementation detail, but I will let you decide that. In any case, I would do the test with `-d` in a separate test function. It looks like an additional, informative thing separate from producing the actual output.

File: python/grass/temporal/temporal_algebra.py [hunk 5]
```diff
@@ -3325,8 +3325,50 @@ def p_expr_t_shift(self, t):
...
+                    # Set new command list for map.
+                    map_new.cmd_list = str(tfuncdict[t[1].upper()])
+
+                    # Append map with updated command list to result list.
+                    resultlist.append(map_new)
+
+            t[0] = resultlist
+
+        if self.debug:
+            print(t[0], t[1], t[2], t[3])
```
Comment (wenzeslaus) [inline python/grass/temporal/temporal_algebra.py:3368]: Seems inconsistent with the rest of prints around here.

File: python/grass/temporal/temporal_algebra.py [hunk 6]
```diff
@@ -3325,8 +3325,50 @@ def p_expr_t_shift(self, t):
...
+        #                     of the time series as a numerical constant
+        #                     for the mapcalculator expression
+
+        """
+        expr : t_var LPAREN NAME COMMA INT RPAREN
+        """
+        if self.run:
+            # Check input stds.
+            maplist = self.check_stds(t[3])
+            resultlist = []
```
Comment (wenzeslaus) [inline python/grass/temporal/temporal_algebra.py:3340]: While the names are consistent with the code around, there is not much reason to be consistent in this case and the resulting name is not a word, so an underscore would probably be nice. One day we will have to apply spell checker and less we have to fix or ignore, the better. Similarly for arith below in tests, although that's function names, not just local variables.

**Review by wenzeslaus — COMMENTED** [review]
The `start_doy(A, 0)` seems as a reasonable solution to the conflict and it could be potentially fixed in the future without breaking the syntax.

No that I'm looking into this more, I miss a description for this PR which would reflect the latest state of the proposed code. The linked issue is quite open on how the problem will be addressed. Our [submitting guidelines](https://github.com/OSGeo/grass/blob/main/doc/development/submitting/general.md#commit-messages) are for one-line commit messages in Subversion times, but the spirit is that the PR should be standalone in terms of information provided, esp. when the problem at hand is complex like this one. I still don't know if I will be able to comment on the actual code, but the having a well-reasoned code will give us more confidence when merging this.

File: temporal/t.rast.algebra/testsuite/test_raster_algebra.py (hunk 4 above)
Comment (wenzeslaus) [reply temporal/t.rast.algebra/testsuite/test_raster_algebra.py]: Separating the two tests revealed that both of them are failing. I guess that's good news and bad news. Anyway, this part is resolved.

File: python/grass/temporal/temporal_algebra.py (hunk 6 above)
Comment (wenzeslaus) [reply python/grass/temporal/temporal_algebra.py]: Thanks. Much better. Resolved.

**Comment by ninsbl** [comment]
Tests were passing locally and the issue was the long mapset name in CI that introduced line breaks in the expression and consequently the tests to fail. We`ll se in an hour or so if this is now fixed. At least now, also long mapset names do not break the local test run...

**Comment by wenzeslaus** [comment]
> ...the long mapset name in CI that introduced line breaks in the expression...

How is the -d output generated? It seems like something which should not be introducing line breaks esp. if one of the use cases for it now is testing of the expressions.

**Comment by ninsbl** [comment]
> How is the -d output generated? It seems like something which should not be introducing line breaks esp. if one of the use cases for it now is testing of the expressions.

It is not obvious. Code is here: https://github.com/OSGeo/grass/blob/a308bdff3ae57353184bc0b15bf105589ddafa78/python/grass/temporal/temporal_raster_base_algebra.py#L355 (I think)

Maybe I abuse the dry-dun flag for testing. In the other cases, it seeems to be only used to assist debugging if a test fails...
However, in my naive thinking it would be most accurate to test if expressions are correct and not only min and max and the number of maps in the time series (as there can be nonsense inbetween)...
But I can drop the test if that way of testing is too fragile...

**Comment by wenzeslaus** [comment]
I think it is about how it is printed (e.g., g.message versus print) rather than how it is constructed. If the string is a real thing used internally, then it makes sense to test it if it is exposed anyway using the -d flag.

**Comment by ninsbl** [comment]
> I think it is about how it is printed (e.g., g.message versus print) rather than how it is constructed. If the string is a real thing used internally, then it makes sense to test it if it is exposed anyway using the -d flag.

Ah, OK. So, the problem is likely that the dry-run uses `pprint` and not simple `print`.

See:
https://github.com/OSGeo/grass/blob/a308bdff3ae57353184bc0b15bf105589ddafa78/temporal/t.rast.algebra/t.rast.algebra.py#L136-L141

`pprint` makes sense as a dictionaries with the expressions that need to be printed can become large. And for the user, I guess the purpose of the dry run is mainly visual inspection in advance of a larger process...

**Comment by ninsbl** [comment]
In principle, line breaks in pprint output are not 100% stable, depending on the length of the mapset name. So, there may be a remaining risk that tests fail due to that. If that disqualifies the printed output from being used in tests, I can remove that test, or maybe pprint could be replaced with `print(json.dumps(pc))`...
I am happy to adjust as preferred. Would be good to finalize the PR...

**Comment by wenzeslaus** [comment]
I can't really comment because I didn't get what the issue was exactly and the particular fix was hard to see from the diff. However, I would hope that the "looks like" function would handle that; that's what it is for. Maybe add some whitespace normalization if needed. `print(json.dumps(pc))` sounds good in general. I don't know if it would be useful here. Would that be as readable as pprint? Would the change create issues in the terms of braking the API? Is full the `format=pprint|json` thing needed?

**Comment by ninsbl** [comment]
OK, now I added handling of line breaks and space, so the tests should be robust with not fully predictable pprint output.

@neteler OK to commit it so it can go into 8.3 or should I hold off?

**Pull request merged**